File: cpp/algorithms/math/matrix/matrix.cpp
#include <alg_std.hpp>
#include <cassert>
using namespace std;

using ll = long long;
const ll MOD = 1000000007LL;

vector<vector<ll>> mat_mul(const vector<vector<ll>>& a, const vector<vector<ll>>& b) {
    if (a.empty() || b.empty()) throw runtime_error("empty matrix");
    int n = (int)a.size(), m = (int)b[0].size(), k = (int)b.size();
    vector<vector<ll>> c(n, vector<ll>(m, 0));
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < m; ++j) {
            ll s = 0;
            for (int t = 0; t < k; ++t) s = (s + a[i][t] * b[t][j]) % MOD;
            c[i][j] = s;
        }
    return c;
}

vector<vector<ll>> mat_pow(vector<vector<ll>> base, long long e) {
    int n = (int)base.size();
    if (n == 0) {
        if (e == 0) return {};
        throw runtime_error("empty matrix");
    }
    vector<vector<ll>> res(n, vector<ll>(n, 0));
    for (int i = 0; i < n; ++i) res[i][i] = 1;
    while (e) {
        if (e & 1) res = mat_mul(res, base);
        base = mat_mul(base, base);
        e >>= 1;
    }
    return res;
}

ll fib(ll n) {
    if (n <= 1) return n;
    vector<vector<ll>> f{{1, 1}, {1, 0}};
    auto p = mat_pow(f, n - 1);
    return p[0][0];
}

int main() {
    assert(fib(10) == 55);
    assert(fib(100) == 687995182);
    auto eye = mat_pow({{1, 1}, {1, 0}}, 0);
    vector<vector<ll>> want{{1, 0}, {0, 1}};
    assert(eye == want);
    try {
        mat_pow({}, 1);
        assert(false);
    } catch (const runtime_error&) {
    }
    cout << "matrix OK" << endl;
    return 0;
}

File: python/algorithms/math/matrix/matrix.py
"""矩阵快速幂：斐波那契。"""

from __future__ import annotations


def mat_mul(a: list[list[int]], b: list[list[int]], mod: int) -> list[list[int]]:
    n = len(a)
    m = len(b[0])
    k = len(b)
    c = [[0] * m for _ in range(n)]
    for i in range(n):
        for j in range(m):
            s = 0
            for t in range(k):
                s = (s + a[i][t] * b[t][j]) % mod
            c[i][j] = s
    return c


def mat_pow(mat: list[list[int]], e: int, mod: int) -> list[list[int]]:
    n = len(mat)
    res = [[1 if i == j else 0 for j in range(n)] for i in range(n)]
    base = [row[:] for row in mat]
    while e:
        if e & 1:
            res = mat_mul(res, base, mod)
        base = mat_mul(base, base, mod)
        e >>= 1
    return res


def fib(n: int, mod: int = 10**9 + 7) -> int:
    if n <= 1:
        return n
    f = [[1, 1], [1, 0]]
    p = mat_pow(f, n - 1, mod)
    return p[0][0]


if __name__ == "__main__":
    mod = 10**9 + 7
    assert fib(10, mod) == 55
    assert fib(100, mod) == 687995182
    print("matrix OK")
